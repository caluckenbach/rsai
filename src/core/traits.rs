use async_trait::async_trait;

use super::{
    error::LlmError,
    types::{BoxFuture, StructuredRequest, StructuredResponse, Tool, ToolRegistry},
};

#[async_trait]
pub trait LlmProvider {
    async fn generate_structured<T>(
        &self,
        request: StructuredRequest,
        tool_registry: Option<&ToolRegistry>,
    ) -> Result<StructuredResponse<T>, LlmError>
    where
        T: serde::de::DeserializeOwned + Send + schemars::JsonSchema;
}

pub trait ToolFunction: Send + Sync {
    fn schema(&self) -> Tool;
    fn execute<'a>(
        &'a self,
        params: serde_json::Value,
    ) -> BoxFuture<'a, Result<serde_json::Value, LlmError>>;
}
