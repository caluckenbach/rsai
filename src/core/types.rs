use crate::core::traits::ToolFunction;
use crate::provider::Provider;
use serde_json::Value;
use std::collections::HashMap;
use std::future::Future;
use std::pin::Pin;
use std::sync::Arc;

#[derive(Debug, Clone, PartialEq)]
pub enum ChatRole {
    System,
    User,
    Assistant,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Message {
    pub role: ChatRole,
    pub content: String,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ToolCall {
    pub id: String,
    pub call_id: String,
    pub name: String,
    pub arguments: Value,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ToolCallResult {
    pub id: String,
    pub tool_call_id: String,
    pub content: serde_json::Value,
}

#[derive(Debug, Clone, PartialEq)]
pub enum ConversationMessage {
    Chat(Message),
    ToolCall(ToolCall),
    ToolCallResult(ToolCallResult),
}

#[derive(Debug, Clone, PartialEq)]
pub struct Tool {
    pub name: String,
    pub description: Option<String>,
    pub parameters: Value,
    pub strict: Option<bool>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum ToolChoice {
    None,
    Auto,
    Required,
    Function { name: String },
}

#[derive(Debug, Clone, PartialEq)]
pub struct StructuredRequest {
    pub model: String,
    pub messages: Vec<ConversationMessage>,
    pub tool_config: Option<ToolConfig>,
    pub generation_config: Option<GenerationConfig>,
}

/// Configuration for tool calling behavior
#[derive(Debug, Clone, PartialEq)]
pub struct ToolConfig {
    /// Available tools for the model to call
    pub tools: Option<Box<[Tool]>>,
    /// Strategy for choosing which tools to call
    pub tool_choice: Option<ToolChoice>,
    /// Whether to allow parallel tool calls (default: true)
    pub parallel_tool_calls: Option<bool>,
}

/// Configuration for text generation parameters
#[derive(Debug, Clone, PartialEq)]
pub struct GenerationConfig {
    /// Maximum number of tokens to generate
    pub max_tokens: Option<u32>,

    /// Sampling temperature
    pub temperature: Option<f32>,

    /// Nucleus sampling parameter (0.0 to 1.0)
    pub top_p: Option<f32>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct StructuredResponse<T> {
    pub content: T,
    pub usage: LanguageModelUsage,
    pub metadata: ResponseMetadata,
}

#[derive(Debug, Clone, PartialEq)]
pub struct LanguageModelUsage {
    pub prompt_tokens: i32,
    pub completion_tokens: i32,
    pub total_tokens: i32,
}

#[derive(Debug, Clone, PartialEq)]
pub struct ResponseMetadata {
    pub provider: Provider,
    pub model: String,
    pub id: String,
}

pub struct ToolRegistry {
    tools: HashMap<String, Arc<dyn ToolFunction>>,
}

impl ToolRegistry {
    pub fn new() -> Self {
        Self {
            tools: HashMap::new(),
        }
    }

    pub fn register(&mut self, tool: Arc<dyn ToolFunction>) {
        let schema = tool.schema();
        self.tools.insert(schema.name, tool);
    }

    pub fn get_schemas(&self) -> Vec<Tool> {
        self.tools.values().map(|tool| tool.schema()).collect()
    }

    pub async fn execute(
        &self,
        tool_call: &ToolCall,
    ) -> Result<serde_json::Value, crate::core::error::LlmError> {
        if let Some(tool) = self.tools.get(&tool_call.name) {
            tool.execute(tool_call.arguments.clone()).await
        } else {
            Err(crate::core::error::LlmError::ToolNotFound(
                tool_call.name.clone(),
            ))
        }
    }
}

pub type BoxFuture<'a, T> = Pin<Box<dyn Future<Output = T> + Send + 'a>>;

pub struct ToolSet {
    pub registry: ToolRegistry,
}

impl ToolSet {
    pub fn tools(&self) -> Vec<Tool> {
        self.registry.get_schemas()
    }
}

impl Default for ToolRegistry {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use async_trait::async_trait;

    use super::*;
    use std::sync::Arc;

    struct ObjectTool;
    #[async_trait]
    impl ToolFunction for ObjectTool {
        fn schema(&self) -> Tool {
            Tool {
                name: "object_tool".to_string(),
                description: Some("Returns an object".to_string()),
                parameters: serde_json::json!({
                    "type": "object",
                    "properties": {},
                    "additionalProperties": false
                }),
                strict: Some(true),
            }
        }

        fn execute<'a>(
            &'a self,
            _params: serde_json::Value,
        ) -> BoxFuture<'a, Result<serde_json::Value, crate::core::error::LlmError>> {
            Box::pin(async move {
                Ok(serde_json::json!({
                    "name": "test",
                    "value":42,
                    "active": true
                }))
            })
        }
    }

    #[tokio::test]
    async fn test_tool_registry_preservers_object_types() {
        let mut registry = ToolRegistry::new();
        registry.register(Arc::new(ObjectTool));

        let tool_call = ToolCall {
            id: "test_Id".to_string(),
            call_id: "call_123".to_string(),
            name: "object_tool".to_string(),
            arguments: serde_json::json!({}),
        };

        let result = registry.execute(&tool_call).await.unwrap();

        // Verify the result is still structured data an not just a string
        assert!(result.is_object());
        assert_eq!(result["name"], "test");
        assert_eq!(result["value"], 42);
        assert_eq!(result["active"], true);
    }
}
