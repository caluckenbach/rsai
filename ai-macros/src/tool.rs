use proc_macro2::TokenStream;
use quote::quote;
use syn::{ItemFn, Result, FnArg, Pat, Type, Attribute};

pub fn tool_impl(attr: TokenStream, item: TokenStream) -> Result<TokenStream> {
    let _ = attr; // Currently unused, could be used for tool configuration
    
    let input = syn::parse2::<ItemFn>(item)?;
    let fn_name = &input.sig.ident;
    let fn_name_str = fn_name.to_string();
    
    // Extract function description and parameter descriptions from doc comments
    let (description, param_descriptions) = extract_doc_comment_and_params(&input.attrs);
    
    // Parse function parameters
    let params = parse_parameters(&input.sig.inputs, &param_descriptions)?;
    
    // Generate JSON schema for parameters
    let schema = generate_parameter_schema(&params)?;
    
    // Generate a function that returns the Tool struct
    let tool_fn_name = quote::format_ident!("{}_tool", fn_name);
    
    // Generate the complete implementation
    let expanded = quote! {
        #input
        
        pub fn #tool_fn_name() -> ai_rs::core::types::Tool {
            ai_rs::core::types::Tool {
                name: #fn_name_str.to_string(),
                description: #description,
                parameters: #schema,
                strict: Some(true),
            }
        }
    };
    
    Ok(expanded)
}

fn extract_doc_comment_and_params(attrs: &[Attribute]) -> (TokenStream, std::collections::HashMap<String, String>) {
    let doc_strings: Vec<String> = attrs
        .iter()
        .filter_map(|attr| {
            if attr.path().is_ident("doc") {
                if let syn::Meta::NameValue(meta) = &attr.meta {
                    if let syn::Expr::Lit(expr_lit) = &meta.value {
                        if let syn::Lit::Str(lit_str) = &expr_lit.lit {
                            return Some(lit_str.value().trim().to_string());
                        }
                    }
                }
            }
            None
        })
        .collect();
    
    let mut description_lines = Vec::new();
    let mut param_descriptions = std::collections::HashMap::new();
    
    for line in doc_strings {
        // Check if this line describes a parameter (format: "param_name: description")
        if let Some(colon_pos) = line.find(':') {
            let param_name = line[..colon_pos].trim();
            let param_desc = line[colon_pos + 1..].trim();
            
            // Only treat it as a parameter description if the param name looks like an identifier
            if param_name.chars().all(|c| c.is_alphanumeric() || c == '_') && !param_name.is_empty() {
                param_descriptions.insert(param_name.to_string(), param_desc.to_string());
                continue;
            }
        }
        
        // Otherwise, it's part of the function description
        description_lines.push(line);
    }
    
    let description = if description_lines.is_empty() {
        quote! { None }
    } else {
        let description = description_lines.join(" ");
        quote! { Some(#description.to_string()) }
    };
    
    (description, param_descriptions)
}

struct Parameter {
    name: String,
    ty: Type,
    description: Option<String>,
    required: bool,
}

fn parse_parameters(
    inputs: &syn::punctuated::Punctuated<FnArg, syn::token::Comma>,
    param_descriptions: &std::collections::HashMap<String, String>,
) -> Result<Vec<Parameter>> {
    let mut params = Vec::new();
    
    for arg in inputs {
        match arg {
            FnArg::Receiver(_) => {
                return Err(syn::Error::new_spanned(arg, "tool functions cannot have self parameter"));
            }
            FnArg::Typed(pat_type) => {
                let name = match &*pat_type.pat {
                    Pat::Ident(pat_ident) => pat_ident.ident.to_string(),
                    _ => return Err(syn::Error::new_spanned(&pat_type.pat, "only simple identifiers are supported for parameters")),
                };
                
                // Get parameter description from docstring parsing
                let description = param_descriptions.get(&name).cloned();
                
                // Check if type is Option<T>
                let (ty, required) = match &*pat_type.ty {
                    Type::Path(type_path) => {
                        let segments = &type_path.path.segments;
                        if segments.len() == 1 && segments[0].ident == "Option" {
                            // Extract inner type from Option<T>
                            if let syn::PathArguments::AngleBracketed(args) = &segments[0].arguments {
                                if let Some(syn::GenericArgument::Type(inner_ty)) = args.args.first() {
                                    (inner_ty.clone(), false)
                                } else {
                                    ((*pat_type.ty).clone(), true)
                                }
                            } else {
                                ((*pat_type.ty).clone(), true)
                            }
                        } else {
                            ((*pat_type.ty).clone(), true)
                        }
                    }
                    _ => ((*pat_type.ty).clone(), true),
                };
                
                params.push(Parameter {
                    name,
                    ty,
                    description,
                    required,
                });
            }
        }
    }
    
    Ok(params)
}


fn generate_parameter_schema(params: &[Parameter]) -> Result<TokenStream> {
    if params.is_empty() {
        return Ok(quote! {
            ::serde_json::json!({
                "type": "object",
                "properties": {},
                "required": [],
                "additionalProperties": false
            })
        });
    }

    let properties: Vec<_> = params.iter().map(|param| {
        let name = &param.name;
        let type_str = type_to_json_type(&param.ty).unwrap_or("string");
        
        if let Some(desc) = &param.description {
            quote! {
                (#name, ::serde_json::json!({
                    "type": #type_str,
                    "description": #desc
                }))
            }
        } else {
            quote! {
                (#name, ::serde_json::json!({
                    "type": #type_str
                }))
            }
        }
    }).collect();

    let required_params: Vec<_> = params.iter()
        .filter(|p| p.required)
        .map(|p| &p.name)
        .collect();
    
    Ok(quote! {
        {
            let mut properties = ::serde_json::Map::new();
            #(
                properties.insert(#properties.0.to_string(), #properties.1);
            )*
            
            ::serde_json::json!({
                "type": "object",
                "properties": properties,
                "required": [#(#required_params),*],
                "additionalProperties": false
            })
        }
    })
}

fn type_to_json_type(ty: &Type) -> Result<&'static str> {
    match ty {
        Type::Path(type_path) => {
            let ident = &type_path.path.segments.last()
                .ok_or_else(|| syn::Error::new_spanned(ty, "empty type path"))?
                .ident;
            
            match ident.to_string().as_str() {
                "String" | "str" => Ok("string"),
                "i8" | "i16" | "i32" | "i64" | "i128" | "isize" |
                "u8" | "u16" | "u32" | "u64" | "u128" | "usize" => Ok("integer"),
                "f32" | "f64" => Ok("number"),
                "bool" => Ok("boolean"),
                "Vec" => Ok("array"),
                _ => Ok("object"), // Default to object for complex types
            }
        }
        _ => Ok("object"),
    }
}