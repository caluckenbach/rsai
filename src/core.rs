pub mod llm;

#[derive(Debug)]
pub enum AIError {
    RequestError(String),
}

pub struct ChatSettings {}

pub struct TextStream {}
