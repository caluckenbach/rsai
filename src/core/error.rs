use thiserror::Error;

#[derive(Debug, Error)]
pub enum LlmError {
    #[error("LLM-Builder error: {0}")]
    Builder(String),

    #[error("Provider configuration error: {0}")]
    ProviderConfiguration(String),

    #[error("Provider error: {message}")]
    Provider {
        message: String,
        #[source]
        source: Option<Box<dyn std::error::Error + Send + Sync>>,
    },

    #[error("Network error: {message}")]
    Network {
        message: String,
        #[source]
        source: Box<dyn std::error::Error + Send + Sync>,
    },

    #[error("API error: {message}")]
    Api {
        message: String,
        status_code: Option<u16>,
        #[source]
        source: Option<Box<dyn std::error::Error + Send + Sync>>,
    },

    #[error("Parse error: {message}")]
    Parse {
        message: String,
        #[source]
        source: Box<dyn std::error::Error + Send + Sync>,
    },

    #[error("Tool execution error: {message}")]
    ToolExecution {
        message: String,
        #[source]
        source: Option<Box<dyn std::error::Error + Send + Sync>>,
    },

    #[error("Tool not found: {0}")]
    ToolNotFound(String),
}
