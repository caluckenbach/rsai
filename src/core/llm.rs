use crate::{
    error::AIError,
    provider::Provider,
    tool::{FunctionDeclaration, Tool},
};
use reqwest::Client;
use serde_json::{Value, json};

pub struct GenerationOptions {
    pub system_prompt: Option<String>,
    pub messages: Option<Vec<Message>>,
    // Add other optional parameters here (temperature, max tokens, etc.)
    pub max_tokens: Option<i32>,
    pub temperature: Option<f32>,
    pub stop_sequences: Option<Vec<String>>,
    pub seed: Option<i32>,
    pub max_retries: Option<i32>,
    // TODO: add abort signal
    // TODO: hide this implementation detail
    pub headers: reqwest::header::HeaderMap,
    // TODO: provider options
}

impl GenerationOptions {
    pub fn new() -> GenerationOptionsBuilder {
        GenerationOptionsBuilder::default()
    }
}

#[derive(Default)]
pub struct GenerationOptionsBuilder {
    system_prompt: Option<String>,
    messages: Option<Vec<Message>>,
    max_tokens: Option<i32>,
    temperature: Option<f32>,
    stop_sequences: Option<Vec<String>>,
    seed: Option<i32>,
    max_retries: Option<i32>,
    headers: reqwest::header::HeaderMap,
}

impl GenerationOptionsBuilder {
    pub fn system_prompt(mut self, prompt: impl Into<String>) -> Self {
        self.system_prompt = Some(prompt.into());
        self
    }

    pub fn messages(mut self, messages: Vec<Message>) -> Self {
        self.messages = Some(messages);
        self
    }

    pub fn max_tokens(mut self, max_tokens: i32) -> Self {
        self.max_tokens = Some(max_tokens);
        self
    }

    pub fn temperature(mut self, temperature: f32) -> Self {
        self.temperature = Some(temperature);
        self
    }

    pub fn stop_sequences(mut self, stop_sequences: Vec<String>) -> Self {
        self.stop_sequences = Some(stop_sequences);
        self
    }

    pub fn seed(mut self, seed: i32) -> Self {
        self.seed = Some(seed);
        self
    }

    pub fn max_retries(mut self, max_retries: i32) -> Self {
        self.max_retries = Some(max_retries);
        self
    }

    pub fn headers(mut self, headers: reqwest::header::HeaderMap) -> Self {
        self.headers = headers;
        self
    }

    pub fn build(self) -> GenerationOptions {
        GenerationOptions {
            system_prompt: self.system_prompt,
            messages: self.messages,
            max_tokens: self.max_tokens,
            temperature: self.temperature,
            stop_sequences: self.stop_sequences,
            seed: self.seed,
            max_retries: self.max_retries,
            headers: self.headers,
        }
    }
}

pub async fn generate_text(
    provider: &dyn Provider,
    prompt: &str,
    options: &GenerationOptions,
) -> Result<String, AIError> {
    let response = provider.generate_text(prompt, options).await?;

    Ok(response)
}

/// Generate content using the LLM with tools
pub async fn generate_content_with_tools(
    client: &Client,
    api_key: &str,
    prompt: &str,
    tools: &[Tool],
) -> Result<Value, AIError> {
    let url = format!(
        "https://generativelanguage.googleapis.com/v1beta/models/gemini-2.0-flash:generateContent?key={}",
        api_key
    );

    // Convert our Tool structs to what Gemini expects
    let function_declarations: Vec<FunctionDeclaration> = tools
        .iter()
        .map(|tool| FunctionDeclaration {
            name: tool.name.clone(),
            description: tool.description.clone(),
            parameters: tool.parameters.clone(),
        })
        .collect();

    let request_body = json!({
        "contents": [{
            "role": "user",
            "parts": [{"text": prompt}]
        }],
        "tools": [{
            "function_declarations": function_declarations
        }],
        "tool_config": {
            "function_calling_config": {
                "mode": "AUTO"
            }
        }
    });

    let response = client
        .post(&url)
        .header("Content-Type", "application/json")
        .json(&request_body)
        .send()
        .await?
        .json::<Value>()
        .await?;

    Ok(response)
}

/// Send function results back to the LLM for a second turn
pub async fn send_function_results(
    client: &Client,
    api_key: &str,
    prompt: &str,
    function_name: &str,
    function_result: &Value,
    tools: &[Tool],
) -> Result<Value, AIError> {
    let url = format!(
        "https://generativelanguage.googleapis.com/v1beta/models/gemini-2.0-flash:generateContent?key={}",
        api_key
    );

    // Convert our Tool structs to what Gemini expects
    let function_declarations: Vec<FunctionDeclaration> = tools
        .iter()
        .map(|tool| FunctionDeclaration {
            name: tool.name.clone(),
            description: tool.description.clone(),
            parameters: tool.parameters.clone(),
        })
        .collect();

    // Create request with conversation history including function result
    let request_body = json!({
        "contents": [
            {
                "role": "user",
                "parts": [{"text": prompt}]
            },
            {
                "role": "model",
                "parts": [{
                    "functionCall": {
                        "name": function_name,
                        "args": {}  // We don't have the original args here
                    }
                }]
            },
            {
                "role": "user",
                "parts": [{
                    "functionResponse": {
                        "name": function_name,
                        "response": {
                            "name": function_name,
                            "content": function_result
                        }
                    }
                }]
            }
        ],
        "tools": [{
            "function_declarations": function_declarations
        }],
        "tool_config": {
            "function_calling_config": {
                "mode": "AUTO"
            }
        }
    });

    let response = client
        .post(&url)
        .header("Content-Type", "application/json")
        .json(&request_body)
        .send()
        .await?
        .json::<Value>()
        .await?;

    Ok(response)
}

/// Extract function call information from LLM response
pub fn extract_function_call(response: &Value) -> Option<(String, Value)> {
    response
        .get("candidates")?
        .as_array()?
        .first()?
        .get("content")?
        .get("parts")?
        .as_array()?
        .first()?
        .get("functionCall")
        .and_then(|function_call| {
            let name = function_call.get("name")?.as_str()?.to_string();
            let args = function_call.get("args")?.clone();
            Some((name, args))
        })
}

/// Extract text response from LLM
pub fn extract_text_response(response: &Value) -> Option<String> {
    response
        .get("candidates")?
        .as_array()?
        .first()?
        .get("content")?
        .get("parts")?
        .as_array()?
        .first()?
        .get("text")?
        .as_str()
        .map(|s| s.to_string())
}
