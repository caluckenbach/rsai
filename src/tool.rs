use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::HashMap;

// Core tool definition
#[derive(Debug, Clone)]
pub struct Tool {
    pub name: String,
    pub description: String,
    pub parameters: Parameters,
    pub execute: fn(HashMap<String, String>) -> Value,
}

// Parameter definitions for tools
#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Parameters {
    #[serde(rename = "type")]
    pub param_type: String,
    pub properties: HashMap<String, Property>,
    pub required: Vec<String>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Property {
    #[serde(rename = "type")]
    pub property_type: String,
    pub description: String,
}

// Function declaration for LLM API
#[derive(Serialize, Deserialize, Debug)]
pub struct FunctionDeclaration {
    pub name: String,
    pub description: String,
    pub parameters: Parameters,
}

// Helper function to create a new tool
pub fn tool(
    name: &str,
    description: &str,
    parameters: HashMap<String, (String, String)>,
    required: Vec<&str>,
    execute: fn(HashMap<String, String>) -> Value,
) -> Tool {
    let mut properties = HashMap::new();
    for (key, (property_type, description)) in parameters {
        properties.insert(
            key.to_string(),
            Property {
                property_type,
                description,
            },
        );
    }

    Tool {
        name: name.to_string(),
        description: description.to_string(),
        parameters: Parameters {
            param_type: "object".to_string(),
            properties,
            required: required.iter().map(|&s| s.to_string()).collect(),
        },
        execute,
    }
}

// Process a function call by finding the appropriate tool and executing it
pub fn process_function_call(
    tools: &[Tool],
    function_name: &str,
    args: &Value,
) -> Option<Value> {
    let tool = tools.iter().find(|t| t.name == function_name)?;

    // Convert args to HashMap<String, String>
    let mut params = HashMap::new();
    if let Some(obj) = args.as_object() {
        for (key, value) in obj {
            if let Some(value_str) = value.as_str() {
                params.insert(key.clone(), value_str.to_string());
            }
        }
    }

    Some((tool.execute)(params))
}