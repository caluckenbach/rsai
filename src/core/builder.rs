use std::{env, marker::PhantomData};

use serde::de::Deserialize;

use crate::provider::openai::{self};

use super::{
    error::LlmError,
    types::{Message, StructuredRequest, Tool, ToolChoice},
};

mod private {
    use crate::core::types::{LlmResponse, StructuredRequest, StructuredResponse};
    use crate::core::error::LlmError;
    use crate::core::traits::LlmProvider;
    use crate::provider::openai::OpenAiClient;
    
    pub struct Init;
    pub struct ProviderSet;
    pub struct ApiKeySet;
    pub struct Configuring;
    pub struct MessagesSet;
    pub struct ToolsSet;

    /// Marker trait for states that can execute complete() requests
    pub trait CompletableState {
        type Output<T>;
        
        async fn execute_request<T>(
            client: &OpenAiClient,
            request: StructuredRequest,
        ) -> Result<Self::Output<T>, LlmError>
        where
            T: serde::de::DeserializeOwned + Send + schemars::JsonSchema;
    }
    
    impl CompletableState for MessagesSet {
        type Output<T> = StructuredResponse<T>;
        
        async fn execute_request<T>(
            client: &OpenAiClient,
            request: StructuredRequest,
        ) -> Result<Self::Output<T>, LlmError>
        where
            T: serde::de::DeserializeOwned + Send + schemars::JsonSchema,
        {
            client.generate_structured(request).await
        }
    }
    
    impl CompletableState for ToolsSet {
        type Output<T> = LlmResponse<T>;
        
        async fn execute_request<T>(
            client: &OpenAiClient,
            request: StructuredRequest,
        ) -> Result<Self::Output<T>, LlmError>
        where
            T: serde::de::DeserializeOwned + Send + schemars::JsonSchema,
        {
            client.generate(request).await
        }
    }
}

/// A type-safe builder for constructing LLM requests using the builder pattern.
/// The builder enforces correct construction order through phantom types.
pub struct LlmBuilder<State> {
    provider: Option<String>,
    model: Option<String>,
    messages: Option<Vec<Message>>,
    api_key: Option<String>,
    tools: Option<Box<[Tool]>>,
    tool_choice: Option<ToolChoice>,
    parallel_tool_calls: Option<bool>,
    _state: PhantomData<State>,
}

impl<State> LlmBuilder<State> {
    pub(crate) fn get_model(&self) -> Option<&str> {
        self.model.as_deref()
    }

    pub(crate) fn set_api_key(&mut self, api_key: &str) {
        self.api_key = Some(api_key.to_string());
    }

    pub(crate) fn get_api_key(&self) -> Option<&str> {
        self.api_key.as_deref()
    }
}

impl LlmBuilder<private::Init> {
    /// Set the provider for the LLM request.
    /// Currently supports "openai".
    pub fn provider(self, provider: &str) -> Result<LlmBuilder<private::ProviderSet>, LlmError> {
        match provider {
            "openai" => Ok(LlmBuilder {
                provider: Some(provider.to_string()),
                model: None,
                messages: None,
                api_key: None,
                tools: None,
                tool_choice: None,
                parallel_tool_calls: None,
                _state: PhantomData,
            }),
            _ => Err(LlmError::Builder("Unsupported Provider".to_string())),
        }
    }
}

/// Configuration for API key source
pub enum ApiKey {
    /// Use the default environment variable for the provider
    Default,
    /// Use a custom API key string
    Custom(String),
}

impl LlmBuilder<private::ProviderSet> {
    /// Set the API key for the provider.
    /// Use `ApiKey::Default` to load from environment variables or `ApiKey::Custom` for a custom key.
    pub fn api_key(self, api_key: ApiKey) -> Result<LlmBuilder<private::ApiKeySet>, LlmError> {
        let key = match api_key {
            ApiKey::Default => match self.provider.as_deref() {
                Some("openai") => env::var("OPENAI_API_KEY").map_err(|_| {
                    LlmError::Builder("Missing OPENAI_API_KEY environment variable".to_string())
                })?,
                _ => {
                    return Err(LlmError::Builder(
                        "Can't load API Key for unsupported Provider".to_string(),
                    ));
                }
            },
            ApiKey::Custom(custom_key) => custom_key,
        };

        Ok(LlmBuilder {
            provider: self.provider,
            model: None,
            messages: None,
            api_key: Some(key),
            tools: None,
            tool_choice: None,
            parallel_tool_calls: None,
            _state: PhantomData,
        })
    }
}

impl LlmBuilder<private::ApiKeySet> {
    /// Set the model to use for the LLM request.
    pub fn model(self, model_id: &str) -> LlmBuilder<private::Configuring> {
        LlmBuilder {
            provider: self.provider,
            model: Some(model_id.to_string()),
            messages: None,
            api_key: self.api_key,
            tools: None,
            tool_choice: None,
            parallel_tool_calls: None,
            _state: PhantomData,
        }
    }
}

impl LlmBuilder<private::Configuring> {
    /// Set the messages for the conversation.
    pub fn messages(self, messages: Vec<Message>) -> LlmBuilder<private::MessagesSet> {
        LlmBuilder {
            provider: self.provider,
            model: self.model,
            messages: Some(messages),
            api_key: self.api_key,
            tools: self.tools,
            tool_choice: self.tool_choice,
            parallel_tool_calls: self.parallel_tool_calls,
            _state: PhantomData,
        }
    }
}

fn validate_builder<State>(
    builder: &LlmBuilder<State>,
) -> Result<(&Vec<Message>, &str, &str), LlmError> {
    builder.api_key.as_ref().ok_or(LlmError::Builder(
        "Missing API key. Make sure to specify an API key.".into(),
    ))?;

    let messages = builder
        .messages
        .as_ref()
        .filter(|messages| !messages.is_empty())
        .ok_or(LlmError::Builder(
            "Missing messages. Make sure to add at least one message.".to_string(),
        ))?;

    let provider = builder.provider.as_ref().ok_or(LlmError::Builder(
        "Missing provider. Make sure to specify a provider.".into(),
    ))?;

    let model = builder.model.as_ref().ok_or(LlmError::Builder(
        "Missing model. Make sure to specify a model.".into(),
    ))?;

    Ok((messages, provider, model))
}

impl<State> LlmBuilder<State>
where
    State: private::CompletableState,
{
    /// Execute the LLM request and return structured output of type T.
    /// The type T must implement Deserialize and JsonSchema for structured output generation as well as
    /// be annotated with `#[schemars(deny_unknown_fields)]`.
    /// Use the `completion_schema` attribute macro to easily define structured output types.
    ///
    /// # Example
    /// ```
    /// use ai::{completion_schema, llm, Message, ChatRole, ApiKey};
    ///
    /// #[completion_schema]
    /// struct Analysis {
    ///     sentiment: String,
    ///     confidence: f32,
    /// }
    ///
    /// let analysis = llm::call()
    ///     .provider("openai")?
    ///     .api_key(ApiKey::Default)?
    ///     .model("gpt-4o-mini")
    ///     .messages(vec![Message {
    ///         role: ChatRole::User,
    ///         content: "Analyze: 'This library is amazing!'".to_string(),
    ///     }])
    ///     .complete::<Analysis>()
    ///     .await?;
    /// ```
    pub async fn complete<T>(mut self) -> Result<State::Output<T>, LlmError>
    where
        T: for<'a> Deserialize<'a> + Send + schemars::JsonSchema,
    {
        let (messages, provider, model) = validate_builder(&self)?;
        let model_string = model.to_string();
        let messages = messages.to_vec();

        match provider {
            "openai" => {
                if self.api_key.is_none() {
                    let api_key = env::var("OPENAI_API_KEY")
                        .map_err(|_| LlmError::Builder("Missing OPENAI_API_KEY.".to_string()))?;
                    self.set_api_key(&api_key);
                }

                let req = StructuredRequest {
                    model: model_string,
                    messages,
                    tools: self.tools.clone(),
                    tool_choice: self.tool_choice.clone(),
                    parallel_tool_calls: self.parallel_tool_calls,
                };
                let client = openai::create_openai_client_from_builder(&self)?;
                State::execute_request(&client, req).await
            }
            _ => todo!(),
        }
    }
}

impl LlmBuilder<private::MessagesSet> {
    /// Set the tools for the LLM request.
    /// This transitions to the ToolsSet state where tool_choice and parallel_tool_calls can be configured.
    ///
    /// By default parallel tool calling is enabled. This can be changed by calling `parallel_tool_calls` with `false`.
    pub fn tools(self, tools: impl Into<Box<[Tool]>>) -> LlmBuilder<private::ToolsSet> {
        LlmBuilder {
            provider: self.provider,
            model: self.model,
            messages: self.messages,
            api_key: self.api_key,
            tools: Some(tools.into()),
            tool_choice: self.tool_choice,
            parallel_tool_calls: Some(true),
            _state: PhantomData,
        }
    }
}

impl LlmBuilder<private::ToolsSet> {
    /// Set the tool choice using a type-safe enum that implements Into<ToolChoice>.
    /// This accepts the generated toolset::Choice enum for compile-time validation.
    pub fn tool_choice<TC>(mut self, choice: TC) -> Self
    where
        TC: Into<ToolChoice>,
    {
        self.tool_choice = Some(choice.into());
        self
    }

    /// Set whether to enable parallel tool calls.
    /// When true, the model can call multiple tools simultaneously.
    pub fn parallel_tool_calls(mut self, enabled: bool) -> Self {
        self.parallel_tool_calls = Some(enabled);
        self
    }
}

/// Module containing the main entry point for building LLM requests
pub mod llm {
    use super::*;

    /// Create a new LLM builder to start constructing a request
    pub fn call() -> LlmBuilder<private::Init> {
        LlmBuilder {
            provider: None,
            model: None,
            messages: None,
            api_key: None,
            tools: None,
            tool_choice: None,
            parallel_tool_calls: None,
            _state: PhantomData,
        }
    }
}
