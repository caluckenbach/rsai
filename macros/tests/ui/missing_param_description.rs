// Use the actual types from the main crate
use rsai::core::{ToolFunction, types::Tool, error::LlmError};
use rsai_macros::tool;

#[tool]
/// Function with missing parameter description
/// param1: First parameter description
/// (param2 description is missing)
fn function_with_missing_desc(param1: String, param2: i32) -> String {
    format!("{} {}", param1, param2)
}

fn main() {}