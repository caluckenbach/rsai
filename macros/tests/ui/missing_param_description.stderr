error: Parameter 'param2' is missing description in docstring. Add: 'param2: description'
 --> tests/ui/missing_param_description.rs:9:1
  |
9 | fn function_with_missing_desc(param1: String, param2: i32) -> String {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused imports: `ToolFunction`, `error::LlmError`, and `types::Tool`
 --> tests/ui/missing_param_description.rs:2:18
  |
2 | use rsai::core::{ToolFunction, types::Tool, error::LlmError};
  |                  ^^^^^^^^^^^^  ^^^^^^^^^^^  ^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default
